source/server/listener_impl.cc

```diff
@@ -82,100 +82,118 @@ bool shouldBindToPort(const envoy::config::listener::v3::Listener& config) {
-      create_socket = true;
+    if (socket_type == Network::Socket::Type::Datagram) {
+      ASSERT(bind_type_ == ListenerComponentFactory::BindType::ReusePort);
     }
   } else {
     ASSERT(local_address_->type() == Network::Address::Type::Pipe);
     // Listeners with Unix domain socket always use shared socket.
-    create_socket = true;
+    // TODO(mattklein123): This should be blocked at the config parsing layer instead of getting
+    // here and disabling reuse port.
+    if (bind_type_ == ListenerComponentFactory::BindType::ReusePort) {
+      bind_type_ = ListenerComponentFactory::BindType::NoReusePort;
+    }
   }
 
-  if (create_socket) {
-    socket_ = createListenSocketAndApplyOptions();
-  }
+  sockets_.push_back(createListenSocketAndApplyOptions(factory, socket_type, options, bind_type_,
+                                                       listener_name, 0));
 
-  if (socket_ && local_address_->ip() && local_address_->ip()->port() == 0) {
-    local_address_ = socket_->addressProvider().localAddress();
+  if (sockets_[0] != nullptr && local_address_->ip() && local_address_->ip()->port() == 0) {
+    local_address_ = sockets_[0]->addressProvider().localAddress();
   }
-  ENVOY_LOG(debug, "Set listener {} socket factory local address to {}", listener_name_,
+  ENVOY_LOG(debug, "Set listener {} socket factory local address to {}", listener_name,
             local_address_->asString());
+
+  // Now create the remainder of the sockets that will be used by the rest of the workers.
+  for (uint32_t i = 1; i < num_sockets; i++) {
+    if (bind_type_ != ListenerComponentFactory::BindType::ReusePort && sockets_[0] != nullptr) {
+      sockets_.push_back(sockets_[0]->duplicate());
+    } else {
+      sockets_.push_back(createListenSocketAndApplyOptions(factory, socket_type, options,
+                                                           bind_type_, listener_name, i));
+    }
+  }
+  ASSERT(sockets_.size() == num_sockets);
 }
 
-Network::SocketSharedPtr ListenSocketFactoryImpl::createListenSocketAndApplyOptions() {
-  // socket might be nullptr depending on factory_ implementation.
-  Network::SocketSharedPtr socket = factory_.createListenSocket(
-      local_address_, socket_type_, options_, {bind_to_port_, !reuse_port_});
+ListenSocketFactoryImpl::ListenSocketFactoryImpl(const ListenSocketFactoryImpl& factory_to_clone)
+    : factory_(factory_to_clone.factory_), local_address_(factory_to_clone.local_address_),
+      socket_type_(factory_to_clone.socket_type_), options_(factory_to_clone.options_),
+      listener_name_(factory_to_clone.listener_name_),
+      tcp_backlog_size_(factory_to_clone.tcp_backlog_size_),
+      bind_type_(factory_to_clone.bind_type_) {
+  for (auto& socket : factory_to_clone.sockets_) {
+    // In the cloning case we always duplicate() the socket. This makes sure that during listener
+    // update/drain we don't lose any incoming connections when using reuse port.
```

Comment:
More comment text to please add, for the next person who looks at this, and doesn't know all the caveats of SO_REUSEPORT:

Linux use of SO_REUSEPORT causes the kernel to allocate a separate queue for each socket on the same address. Incoming connections are immediately assigned to one of these queues. If connections are in the queue when the socket is closed, they are closed/reset, not sent to another queue. So avoid making extra queues in the kernel, even temporarily.